File: cookify_front/src/shared/api/services/MealsService.ts
import { AxiosPromise } from "axios";
import { Meal } from "./../../models/Meal";
import { getWithJwt, postWithJwt, deleteWithJwt } from "./../BaseApi";
import { BASE_API_URL } from "./../../constants/Constants";
import { toJS } from "mobx";
import { DailyMeals } from "./../../models/DailyMeals";
import { AddOrRemoveDailyMealRequest } from "./../../models/AddOrRemoveDailyMealRequest";

interface MealsService {
  getMealsList(jwtToken: string): AxiosPromise<Meal[]>;
  addMealToList(jwtToken: string, request: Meal): AxiosPromise;
  removeMealFromList(jwtToken: string, mealId: string): AxiosPromise;
  updateMealFromList(jwtToken: string, request: Meal): AxiosPromise;
  getDailyMealsList(jwtToken: string): AxiosPromise<DailyMeals[]>;
  addDailyMeal(
    jwtToken: string,
    request: AddOrRemoveDailyMealRequest
  ): AxiosPromise;
  removeDailyMeal(
    jwtToken: string,
    request: AddOrRemoveDailyMealRequest
  ): AxiosPromise;
}

export class DefaultMealsService implements MealsService {
  private getMealsListForUserUrl = "/meals/GetMealsList";
  private addMealToListUrl = "/meals/AddMealToList";
  private removeMealFromListUrl = "/meals/RemoveMealFromList";
  private updateMealFromListUrl = "/meals/UpdateMealFromList";
  private getDailyMealsListUrl = "/meals/GetDailyMealsList";
  private addDailyMealUrl = "/meals/AddDailyMeal";
  private removeDailyMealUrl = "/meals/RemoveDailyMeal";

  getMealsList(jwtToken: string): AxiosPromise<Meal[]> {
    return getWithJwt<Meal[]>(
      BASE_API_URL + this.getMealsListForUserUrl,
      jwtToken
    );
  }
  addMealToList(jwtToken: string, request: Meal): AxiosPromise {
    return postWithJwt<AxiosPromise>(
      BASE_API_URL + this.addMealToListUrl,
      jwtToken,
      request
    );
  }
  removeMealFromList(jwtToken: string, mealId: string): AxiosPromise {
    return deleteWithJwt<AxiosPromise>(
      BASE_API_URL + this.removeMealFromListUrl,
      jwtToken,
      { MealId: mealId }
    );
  }
  updateMealFromList(jwtToken: string, request: Meal): AxiosPromise {
    return postWithJwt<AxiosPromise>(
      BASE_API_URL + this.updateMealFromListUrl,
      jwtToken,
      toJS(request)
    );
  }
  getDailyMealsList(jwtToken: string): AxiosPromise<DailyMeals[]> {
    return getWithJwt<DailyMeals[]>(
      BASE_API_URL + this.getDailyMealsListUrl,
      jwtToken
    );
  }
  addDailyMeal(
    jwtToken: string,
    request: AddOrRemoveDailyMealRequest
  ): AxiosPromise {
    return postWithJwt<AxiosPromise>(
      BASE_API_URL + this.addDailyMealUrl,
      jwtToken,
      request
    );
  }
  removeDailyMeal(
    jwtToken: string,
    request: AddOrRemoveDailyMealRequest
  ): AxiosPromise {
    return deleteWithJwt<AxiosPromise>(
      BASE_API_URL + this.removeMealFromListUrl,
      jwtToken,
      request
    );
  }
}

export default new DefaultMealsService();

File: Cookify.API/Services/Meals/MealsService.cs
using Cookify.API.Models.Common;
using Cookify.API.Models.Repository;
using Cookify.API.Models.Requests;
using Cookify.API.Repositories.Users;
using Cookify.API.Services.Shopping;
using Microsoft.Extensions.Logging;
using MongoDB.Bson;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Cookify.API.Services.Meals
{
    public class MealsService : IMealsService
    {
        private readonly ILogger _logger;
        private readonly IUserRepository _userRepository;

        public MealsService(ILogger<ShoppingService> logger, IUserRepository userRepository)
        {
            _logger = logger;
            _userRepository = userRepository;
        }

        public ServiceResponse<List<Meal>> GetMealsList(string userId)
        {
            try
            {
                var result = _userRepository.GetMealsList(userId);

                if (result == null)
                {
                    return ServiceResponse<List<Meal>>.Failed();
                }

                return ServiceResponse<List<Meal>>.Succeeded(result);
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse<List<Meal>>.Failed();
            }

        }

        public ServiceResponse AddMealToList(string userId, AddMealRequest addMealRequest)
        {
            try
            {
                var user = _userRepository.GetWhere(x => x.Id == userId);

                var mealToAdd = new Meal
                {
                    Id = ObjectId.GenerateNewId().ToString(),
                    Name = addMealRequest.Name,
                    Ingredients = addMealRequest.Ingredients,
                    Recipe = addMealRequest.Recipe,
                    AdditionalInfo = addMealRequest.AdditionalInfo
                };

                var result = _userRepository.AddMealToList(user.Id, mealToAdd);

                if (result == null)
                {
                    return ServiceResponse.Failed();
                }

                if (result.Count() == user.MealsList.Count + 1)
                {
                    return ServiceResponse.Succeeded();
                }
                else
                {
                    return ServiceResponse.Failed();
                }
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse.Failed();
            }
        }

        public ServiceResponse RemoveMealFromList(string userId, string mealId)
        {
            try
            {
                var user = _userRepository.GetWhere(x => x.Id == userId);

                var result = _userRepository.RemoveMealFromList(user.Id, mealId);

                if (result == null)
                {
                    return ServiceResponse.Failed();
                }

                if (result.Count() == user.MealsList.Count - 1)
                {
                    return ServiceResponse.Succeeded();
                }
                else
                {
                    return ServiceResponse.Failed();
                }
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse.Failed();
            }
        }

        public ServiceResponse UpdateMealFromList(string userId, UpdateMealRequest request)
        {
            try
            {

                var mealToUpdate = new Meal
                {
                    Id = request.Id,
                    Name = request.Name,
                    Ingredients = request.Ingredients,
                    Recipe = request.Recipe,
                    AdditionalInfo = request.AdditionalInfo
                };

                var result = _userRepository.UpdateMealFromList(userId, mealToUpdate);

                if (result == null)
                {
                    return ServiceResponse.Failed();
                }

                if (result.Name == request.Name
                    && result.Recipe == request.Recipe
                    && result.AdditionalInfo == request.AdditionalInfo
                    && result.Ingredients.SequenceEqual(request.Ingredients))
                {
                    return ServiceResponse.Succeeded();
                }
                else
                {
                    return ServiceResponse.Failed();
                }
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse.Failed();
            }
        }
        public ServiceResponse<List<DailyMeals>> GetDailyMealsList(string userId)
        {
            try
            {
                var result = _userRepository.GetDailyMealsList(userId);

                if (result == null)
                {
                    return ServiceResponse<List<DailyMeals>>.Failed();
                }

                return ServiceResponse<List<DailyMeals>>.Succeeded(result);
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse<List<DailyMeals>>.Failed();
            }

        }

        public ServiceResponse AddDailyMeal(string userId, AddOrRemoveDailyMealRequest request)
        {
            try
            {
                var result = _userRepository.AddOrUpdateDailyMeal(userId, request);

                if (result == null)
                {
                    return ServiceResponse.Failed();
                }

                return ServiceResponse.Succeeded();
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse.Failed();
            }
        }
        public ServiceResponse RemoveDailyMeal(string userId, AddOrRemoveDailyMealRequest request)
        {
            try
            {
                var result = _userRepository.RemoveDailyMeal(userId, request);

                if (result == null)
                {
                    return ServiceResponse.Failed();
                }

                return ServiceResponse.Succeeded();
            }
            catch (Exception ex)
            {
                _logger.LogCritical(ex.ToString());
                return ServiceResponse.Failed();
            }
        }

    }
}
